ambry-store/src/main/java/com.github.ambry.store/StoreFindToken.java

```diff
@@ -245,52 +303,105 @@ public long getBytesRead() {
-    if (type != Type.Uninitialized) {
-      // add incarnationId
-      bufWrap.putInt(incarnationIdBytes.length);
-      bufWrap.put(incarnationIdBytes);
-      // add sessionId
-      bufWrap.putInt(sessionIdBytes.length);
-      bufWrap.put(sessionIdBytes);
-      // add offset
-      bufWrap.put(offsetBytes);
-      if (type == Type.JournalBased) {
-        bufWrap.put(getInclusive() ? (byte) 1 : (byte) 0);
-      } else if (type == Type.IndexBased) {
+    byte[] buf = null;
+    switch (version) {
+      case VERSION_0:
+        int OFFSET_SIZE = 8;
+        int START_OFFSET_SIZE = 8;
+        byte[] sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        byte[] storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+        int size = VERSION_SIZE + SESSION_ID_LENGTH_SIZE + sessionIdBytes.length +
+            OFFSET_SIZE + START_OFFSET_SIZE + storeKeyBytes.length;
+        buf = new byte[size];
+        ByteBuffer bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(version);
+        // add sessionId
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
+        // add offset for journal based token
+        bufWrap.putLong((type == Type.JournalBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
+            : UNINITIALIZED_OFFSET);
+        // add index start offset for Index based token
+        bufWrap.putLong((type == Type.IndexBased) ? (offset != null ? offset.getOffset() : UNINITIALIZED_OFFSET)
+            : UNINITIALIZED_OFFSET);
+        // add storekey
         bufWrap.put(storeKeyBytes);
-      }
+        break;
+      case VERSION_1:
+        byte[] offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
+        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+        size = VERSION_SIZE + SESSION_ID_LENGTH_SIZE + sessionIdBytes.length + TYPE_SIZE + offsetBytes.length
+            + storeKeyBytes.length;
+        buf = new byte[size];
+        bufWrap = ByteBuffer.wrap(buf);
+        // add version
+        bufWrap.putShort(VERSION_1);
+        // add sessionId
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
+        // add type
+        bufWrap.putShort((short) type.ordinal());
+        // add offset
+        bufWrap.put(offsetBytes);
+        // add StoreKey
+        bufWrap.put(storeKeyBytes);
+        break;
+      case VERSION_2:
+        offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
+        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+        if (type != Type.Uninitialized && incarnationId == null) {
+          throw new IllegalStateException("IncarnationId cannot be null for Journal or Index based token ");
```

Comment:
yes, we can't do it easily. Check for serDeTest() in StoreFindTokenTest(). We do create tokens in old format and we have a private method to return the serialized form for the given version. With the extra chuck for null incarnationId, we can do that. Have to introduce another method which takes in all arguments and the required version to return the serialized form of it.